CARDS-1669: Add support for exporting data as csv

veronikaslc opened the pull request:
Added a functionality to the DataProcessor that serializes all forms associated with a questionnaire
To test:
- Create some forms
- go to `http: localhost:8080//Questionnaires/<Questionnaire name>.csv` (or: in the `Administration > Questionnaires`,  click on a questionnaire name, then remove from that url `/content.html/admin` and add `.csv` at the end)
Expected: 
- A CSV file will be offered to download with the followding columns:

-     <Subject type> ID (e.g. “Patient ID” for any cards4care forms),
-     Created (jcr:created time stamp in ISO datetime format without timezone yyyy-MM-ddTHH:mm:ss), 
-     All questions, one per column, in the order of their appearance in the questionnaire; the question text will be the column label. Section labels should not be exported at this time

Commit:
CARDS-1669: Add support for exporting data as csv
Added a functionality to the DataProcessor that serializes all forms associated with a questionnaire

Commit:
CARDS-1669: Add support for exporting data as csv
Added a CSVProcessor to serialize the forms associated with each questionnaire to CSV

Commit:
CARDS-1669: Add support for exporting data as csv
Added Logger t the CSV processor class

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/CSVShell.java:
```diff
@@ -0,0 +1,39 @@
+ *
+ * @version $Id$
+ */
+public class CSVShell
```
Rename to CSVString

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,261 @@
+        // The proper serialization depends on "deep", "dereference" and "labels", but we may allow other JSON
+        // processors to be enabled/disabled to further customize the data, so we also append the original selectors
+        final String processedPath = resource.getPath()
+            + resource.getResourceMetadata().getResolutionPathInfo() + ".deep.dereference.labels";
```
```suggestion
            + resource.getResourceMetadata().getResolutionPathInfo() + ".deep.dereference.-labels";
```
Deactivate labels by default.

marta- replied:
To be postponed, not necessary in the first iterration.

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,261 @@
+            // corresponding answer strings to form scv
+            final Map<String, List<String>> csvData = new LinkedHashMap<>();
+            csvData.put("Patient ID", new ArrayList<>());
+            csvData.put("Created", new ArrayList<>());
```
`Patient ID` and `Created` are repeated 4 times in this file. Any repeated strings should be declared as constants.

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,261 @@
+            // CSV data strings aggregator where the key is Question uuid that maps to the list of
+            // corresponding answer strings to form scv
+            final Map<String, List<String>> csvData = new LinkedHashMap<>();
+            csvData.put("Patient ID", new ArrayList<>());
```
The subject won't always be a patient, sometimes it's a tumor, visit, etc. The specs say to use the actual subject type label in the column header. More specifically:
* Check `subjectTypes` specified in the questionnaire definition
* Add column for each type and for each of its parents, sorted by the type's default order, and labeled by `<type.label> ID`. For example:
  * in **proms**, `GAD7` has  `subjectTypes=[Visit]`, so we will have TWO subject id columns: `Patient ID` and `Visit ID`, displaying the parent patient subject's `identifier` and the visit subject's `identifier` respectively.
  * in **proms**, `Patient information` has `subjectTypes=[Patient]`, so we will have ONE subject id columns: `Patient ID`
  * in **lfs**, `Radiotherapy` has `subjectTypes=[Patient, Tumor]`, so we will have TWO subject id columns: `Patient ID` and `Tumor ID`; `Tumor ID` will be blank for any form associated with a patient subject rather than a tumor subject
  * in **lfs**, `SomaticVariants` has `subjectTypes=[Tumor, TumorRegion]`, so we will have THREE subject id columns: `Patient ID`, `Tumor ID`, and `Tumor region ID`; `Tumor region ID` will be blank for any form associated with a Tumor subject rather than a Region subject.

While these hierarchical structure details don't apply directly to cards4care (which is our immediate goal), they should be implemented for broader use in the platform.

If we choose to merge without these to save time, I suggest we move the functionality in the cards4care module until we're ready to expand it.

marta- replied:
To be postponed, not necessary in the first iteration.

veronikaslc replied:
Done

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,261 @@
+            String uuid = nodeJson.getJsonObject("question").getString("jcr:uuid");
+            List<String> answerColumn = csvData.get(uuid);
+
+            final JsonValue value = nodeJson.get("value");
+            if (value == null) {
+                // Fallback to "displayedValue"
+                final JsonValue displayedValue = nodeJson.get("displayedValue");
+                if (displayedValue == null) {
+                    answerColumn.add("");
+                } else {
+                    answerColumn.add(((JsonObject) displayedValue).getString("displayedValue"));
+                }
```
Displaying values or labels in the CSV should both be possible, with values as the default.

To achieve this:
* Serialize forms using the `-labels` processor by default, which will remove the `displayedValue` (see previous comment on line 66).
* The user will have the option to add the `labels` processor to the URL (`/Questionnaires/GAD7.labels.csv`) which will turn labels back on, and serialize forms with `displayedValue` included
* In this block, reverse the roles of `value` and `displayedValue`: export `displayedValue` if it exists, otherwise export `value`

marta- replied:
to be postponed, not necessary in the first iteration.

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,261 @@
+                    answerColumn.add(((JsonObject) displayedValue).getString("displayedValue"));
+                }
+            } else if ("cards:PedigreeAnswer".equals(nodeType) || "cards:FileAnswer".equals(nodeType)) {
+                answerColumn.add("yes");
```
For Files, we should display the file name instead of a boolean
```suggestion
            } else if ("cards:PedigreeAnswer".equals(nodeType)) {
                answerColumn.add("yes");
```

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,261 @@
+                    answerColumn.add(((JsonString) value).getString());
+                } else {
+                    answerColumn.add(value.toString());
+                }
```
This block ignores the case when we're supposed to export the `displayedValue` and the `displayedValue` is an array.

I think the logic should be:
* get the `value`
* is the value `null`? display `""`
* is the question type Pedigree? display "yes" (we already know the value isn't null)
* is there a `displayedValue`? switch to it from this point on, otherwise continue with `value`
* is it an array? join before displaying
* otherwise display it

marta- replied:
This flow, with the exception of `displayValue` use, is already covered by the two outstanding comments above for this method.

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireCSVServlet.java:
```diff
@@ -0,0 +1,58 @@
+        throws IOException
+    {
+        Resource questionnaire = request.getResource();
+        CSVShell csv = questionnaire.adaptTo(CSVShell.class);
```
This should eliminate the need to specify `.data` before `.csv` in the URL.

```suggestion
        final String csvPath = questionnaire.getPath() + ".data" + questionnaire.getResourceMetadata().getResolutionPathInfo();
        CSVShell csv = questionnaire.getResourceResolver().resolve(csvPath).adaptTo(CSVShell.class);
```

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,261 @@
+
+            final JsonValue value = nodeJson.get("value");
+            if (value == null) {
+                // Fallback to "displayedValue"
+                final JsonValue displayedValue = nodeJson.get("displayedValue");
+                if (displayedValue == null) {
+                    answerColumn.add("");
+                } else {
+                    answerColumn.add(((JsonObject) displayedValue).getString("displayedValue"));
+                }
```
Let's ignore the `displayedValue` in the first iteration and just work with `value`.
```suggestion
                answerColumn.add("");
```

Commit:
CARDS-1669: Add support for exporting data as csv
Fixed generation of recurrent sections

Commit:
CARDS-1669: Add support for exporting data as csv
Fixed export for array for long values

Commit:
CARDS-1669: Add support for exporting data as csv
Export vocabulary ID instead of path

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,365 @@
+     * @param answerSection section json object
+     * @param recurrentSections list of already added recurrent sections
+     */
+    private void recordOccurentSection(final JsonObject answerSection, List<String> recurrentSections)
```
```suggestion
    private void recordRecurrentSection(final JsonObject answerSection, List<String> recurrentSections)
```

Commit:
CARDS-1669: Add support for exporting data as csv
Renamed the function

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,365 @@
+        if ("cards:Section".equals(nodeType)) {
+            final String displayMode = getDisplayMode("section", nodeJson);
+            if (displayMode != null && !"default".equals(displayMode)) {
+                // Do not output summary, headers or footers sections
```
Are we sure about this? In proms, the score is in a header section, and I think we would want the score to be included in the CSV.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/DataProcessor.java:
```diff
@@ -129,11 +134,15 @@ public void leave(Node node, JsonObjectBuilder json, Function<Node, JsonValue> s
         try {
-            final Node questionnaire = form.adaptTo(Node.class).getProperty("questionnaire").getNode();
-            final String questionnaireTitle = questionnaire.getProperty("title").getString();
+            String questionnaireTitle = "data";
```
Just `data` may collide with one of the questionnaire's sections or questions. I would use `@data`, since we use the `@` prefix for other custom generated properties.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,365 @@
+        }
+
+        //Empty csvData for future form
+        level = 0;
```
This is useless, since this is a local variable that will be erased when the function returns.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,365 @@
+        return null;
+    }
+
+    private void getSubjectTypes(JsonArray subjectTypesArray, Map<String, Map<Integer, String>> csvData,
```
This outputs the subjects as `tumor region, tumor, patient`, but it would make more sense to have them in the other direction.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,365 @@
+    private String processQuestionnaire(JsonObject questionnaire)
+    {
+        try {
+            final StringWriter stringWriter = new StringWriter();
```
A `StringWriter` is a synchronized class, meaning that it has extra checks to make sure multi-threaded access to it is safe, which adds overhead and makes it slower. Since we don't need thread-safety here, a `StringBuilder` is better suited.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,365 @@
+            int newLevel = level;
+            final String uuid = nodeJson.getJsonObject("section").getString(UUID_PROP);
+            if (recurrentSections.contains(uuid)) {
+                newLevel++;
```
This can only happen once, since `contains` is a simple indicator if the section occurred at least once before. The outcome is that a section repeated more than twice will only output the first and the last instance, none of the ones in the middle.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireCSVServlet.java:
```diff
@@ -0,0 +1,60 @@
+        final String csvPath = questionnaire.getPath() + ".data"
+            + questionnaire.getResourceMetadata().getResolutionPathInfo();
+        CSVString csv = questionnaire.getResourceResolver().resolve(csvPath).adaptTo(CSVString.class);
+        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX");
```
This is too much detail, we don't really want the milliseconds and the timezone in there. Also `:` is not a valid windows character, so it should be replaced with something else. `yyyy-MM-dd'T'HHmm` is enough for the date.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,365 @@
+                                             .get();
+
+        // Assemble CSV by rows
+        for (int i = 0; i < level + 1; i++) {
```
`i < level + 1` is the same as `i <= level`.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,365 @@
+            processFormSection(nodeJson, csvData, recurrentSections, newLevel);
+        } else if (nodeType.startsWith("cards:") && nodeType.endsWith("Answer")) {
+            final String uuid = nodeJson.getJsonObject("question").getString(UUID_PROP);
+            final Map<Integer, String> answerColumn = csvData.get(uuid);
```
If this is in a skipped section/question (i.e. hidden question or summary section), `answerColumn` will be `null`, which will trigger a NPE on next line. A `!= null` check is needed.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,365 @@
+    private String getDisplayMode(final String element, final JsonObject answerElementJson)
+    {
+        try {
+            return ((JsonString) answerElementJson.getValue("/" + element + "/displayMode")).getString();
```
This always returns `null` because the path you are using is valid when `answerElementJson` is an actual answer or answer section, but you only call it with a question or section definition, where the display mode is a direct property of the element itself.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/labels/ResourceCSVProcessor.java:
```diff
@@ -0,0 +1,55 @@
+/*
```
This should be in `modules/utils/src/main/java/io/uhndata/cards/serialize/spi/ResourceCSVProcessor.java`, next to the other resource processor interfaces.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/ResourceToCSVAdapterFactory.java:
```diff
@@ -0,0 +1,65 @@
+/*
```
This should be in `modules/utils/src/main/java/io/uhndata/cards/serialize/ResourceToCSVAdapterFactory.java`, next to the other adapter factories.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/CSVString.java:
```diff
@@ -0,0 +1,39 @@
+        this.data = info;
+    }
+
+    public String getData()
```
This should just be `toString`.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,365 @@
+                if (answer == null) {
+                    answer = "";
+                }
+                row.add(answer);
```
The issue description said that the subject identifier and created date should be repeated on every extra row, but this isn't happening here.

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Move classes in the right place

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Javadoc and codestyle improvements

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
In the questionnaire JSON, put the data in `@data` instead of `data` to avoid collisions with actual questionnaire sections or questions

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Bugfix: only the first and last instances of recurrent sections are exported

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Bugfix: NPE when a question is hidden

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Bugfix: wrong way of getting the display mode of questions/sections

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Simplify output file name

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Better method for converting a CSVString into a plain String

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Don't ignore non-default sections

sdumitriu committed:
[misc] Code optimizations

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Replace StringWriter with StringBuilder for increased performance

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Add an identifier column to link rows belonging to the same form

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,362 @@
+            .map(JsonValue::asJsonObject)
+            .filter(value -> value.containsKey(PRIMARY_TYPE_PROP)
+                && "cards:SubjectType".equals(value.getString(PRIMARY_TYPE_PROP)))
+            .forEach(value -> processSubjectType(value, csvData, columns));
```
This does not include types **higher up** in the hierarchy, but **lower**, e.g. in LFS mode the Patient Information questionnaire requires patients, but this code will include all 3 types, Patient, Tumor, and Tumor Region. Also, this does not consider duplicates, so Chemotherapy, which requires either Patient or Tumor, will output the following columns: `Patient ID, Tumor ID, Tumor Region ID, Tumor ID, Tumor Region ID`.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,362 @@
+            .map(JsonValue::asJsonObject)
+            .filter(value -> value.containsKey(PRIMARY_TYPE_PROP)
+                && "cards:Subject".equals(value.getString(PRIMARY_TYPE_PROP)))
+            .forEach(value -> processFormsSubjects(value, csvData));
```
Just like for subject types in the header, this does not look only at parent subjects, but at children too, so a form that applies to a Patient will also list a Tumor identifier if the patient happens to have a Tumor child, even though the form does not apply to it. It should look explicitly at `subjectJson.getJsonObject("parents")`.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,362 @@
+            List<String> columns = new ArrayList<>();
+
+            // Fetch the subject types expected to be for the questionnaire
+            if (questionnaire.containsKey("requiredSubjectTypes")) {
```
If the questionnaire does not specify required subject types, then it applies to any subject type, but the CSV will not list any subjects at all. An `else` branch is needed, that will look for all subject types in the system and add a column for each of them, sorted by their `cards:defaultOrder` values.

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Bugfix: if a questionnaire does not require specific subject types, all subject types should be included

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Bugfix: subject type columns may be duplicated and out of order

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Bugfix: unrelated child subjects may be included for a form

marta- committed:
CARDS-1669: Add support for exporting data as csv
Turn off the `labels` processor by default, and export the `displayedValue` instead of the `value` if `labels` is specified in the csv export path (e.g. `/Questionnaires/Q.labels.csv`)

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,376 @@
+
+        // As we collect csv data to the csvData we record recurrent sections we already added
+        // to place all subsequent on a new lines
+        final List<String> recurrentSections = new ArrayList<>();
```
Where is `recurrentSections` used? I can't seem to find any code that reads this list.

sdumitriu replied:
You are right, it is no longer used anywhere. It was used to decide when to increase the level of the current instance, but wrongly.

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Remove unused code

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,350 @@
+     */
+    private void processFormSection(final JsonObject answSectionJson, Map<String, Map<Integer, String>> csvData)
+    {
+        answSectionJson.values().stream()
+            .filter(value -> ValueType.OBJECT.equals(value.getValueType()))
+            .map(JsonValue::asJsonObject)
+            .filter(value -> value.containsKey(PRIMARY_TYPE_PROP))
+            .forEach(value -> processFormElement(value, csvData));
```
Could this work as a quick fix for the display of nested recurring sections?
* instead of exporting form elements in the order of their occurrence, export child sections first, and return how many rows each took
* obtain the max of the number of rows occupied by each child section
* then process child questions, and set their row index to skip over this max computed above instead of the assuming it is the size of the already computed list for this column (line 299)

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Fixed bad javadoc

Commit:
CARDS-1669: Add support for exporting data as csv
Added support of export for nested recurrent sections

veronikaslc commented:
Rebased on the latest dev. Added support of proper export of nested recurring sections.

sashaandjic commented:
I created 2 Physical Health Assessment forms for one patient and ended up having each form appear in two rows.
![image](https://user-images.githubusercontent.com/23243244/161306737-e2255101-3f07-461c-a4be-98b56a595626.png)

Commit:
CARDS-1669: Add support for exporting data as csv
Fixed NPE for the requited subject cases

Commit:
CARDS-1669: Add support for exporting data as csv
Fixed csv export for sibling sections

Commit:
CARDS-1669: Add support for exporting data as csv
Fixed printing of special characters

sashaandjic reviewed (approved):
tested - looks good
QUESTION:
Can we/Should we fix ** formatting and displays text bolded; also the links can be prettified? See the attached screenshots:
![image](https://user-images.githubusercontent.com/23243244/162032815-0098ac11-7e37-4aa8-86b6-cf71b91d8f81.png)
![image](https://user-images.githubusercontent.com/23243244/162032990-98438737-fd0f-40b8-8bad-8ef3b890e581.png)

![image](https://user-images.githubusercontent.com/23243244/162032761-fa638e56-94e3-4de1-90fe-24e72fa2b6d0.png)

![image](https://user-images.githubusercontent.com/23243244/162032702-6848c880-dc4d-4c1c-97aa-4b30f395bea2.png)

![image](https://user-images.githubusercontent.com/23243244/162032658-7fcd3c00-77c8-4ee3-9b1a-63a06640536f.png)

sashaandjic commented:
Completely unrelated question: Can we use more user friendly values in the field `Others using tobacco at home`?
e.g. parents, children, etc

veronikaslc commented:
rebased on the latest dev

sashaandjic reviewed (approved):
tested

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Use the SlingSafeMethodsServlet as the base class, since only GET support is needed

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,414 @@
+    {
+        final String nodeType = nodeJson.getString(PRIMARY_TYPE_PROP);
+        if ("cards:Section".equals(nodeType)) {
+            getDisplayMode(nodeJson);
```
This does nothing. Do we want to skip certain types of sections, or remove this line?

sdumitriu replied:
@marta- ^

marta- replied:
you can remove it

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/TreeGraph.java:
```diff
@@ -0,0 +1,132 @@
+import java.util.List;
+import java.util.Map;
+
+public class TreeGraph<T>
```
The type parameter seems to be unused. We only ever put strings in it, and we use it untyped, which triggers lots of warnings. Do we ever expect to have something other than strings in the graph, or can the type be removed?

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,414 @@
+
+            if (node.isSection()) {
+                isRecurring = node.isRecurrentSection();
+                isSameSection = sectionId == node.getUuid();
```
This assumes that instances of the same repeatable section are close together, but this is not always the case. The order is the order in which sections have been added, so having two different repeatable sections (sA and sB), each with two instances (sA1, sA2, sB1, sB2), added not two of the first then two of the second, but sA1 sB1 sA2 sB2, only sA2 and sB2 will be present in the output.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/QuestionnaireToCsvProcessor.java:
```diff
@@ -0,0 +1,414 @@
+     * @param csvData data aggregator
+     * @param formGraph tree structure
+     */
+    private void processFormSection(final JsonObject answSectionJson, Map<String, Map<Integer, String>> csvData,
```
`csvData` is passed around between `processFormSection` and `processFormElement`, but is never used. Should be removed.

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Bugfix: repeatable sections may be lost in the export
Also:
- use boolean instead of Boolean
- improve comments
- closing a writer should happen before reading its contents to make sure no data is lost in the buffer
- refactored code to pass checkstyle
- removed type parameter from TreeGraph since only one type is ever needed

sdumitriu committed:
CARDS-1669: Add support for exporting data as csv
Codestyle improvements and simplification

sashaandjic commented:
tested

sashaandjic reviewed (approved):
tested
